packages/edit-navigation/src/store/actions.js

```diff
@@ -49,171 +39,366 @@ export function setSelectedMenuId( menuId ) {
  * @param {Object} post A navigation post to process
  * @return {Function} An action creator
  */
-export const createMissingMenuItems = serializeProcessing( function* ( post ) {
+export const createMissingMenuItems = ( post ) => async ( {
+	dispatch,
+	registry,
+} ) => {
 	const menuId = post.meta.menuId;
+	const lock = await registry
+		.dispatch( coreDataStore )
+		.__unstableAcquireStoreLock( STORE_NAME, [ 'savingMenu' ], {
+			exclusive: false,
+		} );
+	try {
+		// Ensure all the menu items are available before we start creating placeholders.
+		await registry
+			.resolveSelect( coreDataStore )
+			.getMenuItems( { menus: menuId, per_page: -1 } );
 
-	const mapping = yield getMenuItemToClientIdMapping( post.id );
-	const clientIdToMenuId = invert( mapping );
-
-	const stack = [ post.blocks[ 0 ] ];
-	while ( stack.length ) {
-		const block = stack.pop();
-		if ( ! ( block.clientId in clientIdToMenuId ) ) {
-			const menuItem = yield apiFetch( {
-				path: `/__experimental/menu-items`,
-				method: 'POST',
-				data: {
-					title: 'Placeholder',
-					url: 'Placeholder',
-					menu_order: 1,
-				},
-			} );
+		const menuItemIdToBlockId = await dispatch(
+			getEntityRecordIdToBlockIdMapping( post.id )
+		);
+		const knownBlockIds = new Set( Object.values( menuItemIdToBlockId ) );
 
-			mapping[ menuItem.id ] = block.clientId;
-			const menuItems = yield resolveMenuItems( menuId );
-			yield dispatch(
-				'core',
-				'receiveEntityRecords',
-				'root',
-				'menuItem',
-				[ ...menuItems, menuItem ],
-				menuItemsQuery( menuId ),
-				false
-			);
+		const blocks = blocksTreeToFlatList( post.blocks[ 0 ].innerBlocks );
+		for ( const { block } of blocks ) {
+			if ( block.name !== 'core/navigation-link' ) {
+				continue;
+			}
+			if ( ! knownBlockIds.has( block.clientId ) ) {
+				const menuItem = await dispatch(
+					createPlaceholderMenuItem( menuId )
+				);
+				menuItemIdToBlockId[ menuItem.id ] = block.clientId;
+			}
 		}
-		stack.push( ...block.innerBlocks );
+
+		dispatch( {
+			type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
+			postId: post.id,
+			mapping: menuItemIdToBlockId,
+		} );
+	} finally {
+		registry.dispatch( coreDataStore ).__unstableReleaseStoreLock( lock );
 	}
+};
 
-	yield {
-		type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
-		postId: post.id,
-		mapping,
-	};
-} );
+/**
+ * Creates a single placeholder menu item.
+ * Requests POST /wp/v2/menu-items once for every menu item created.
+ *
+ * @param {number} menuId Menu id to embed the placeholder in.
+ * @return {Function} An action creator
+ */
+export const createPlaceholderMenuItem = ( menuId ) => async ( {
+	registry,
+} ) => {
+	const existingMenuItems = await registry
+		.select( coreDataStore )
+		.getMenuItems( { menus: menuId, per_page: -1 } );
+
+	const createdMenuItem = await apiFetch( {
+		path: `/__experimental/menu-items`,
+		method: 'POST',
+		data: {
+			title: 'Placeholder',
+			url: 'Placeholder',
```

Comment:
I don't love these values. How about these values. 

```suggestion
			title: __( 'Menu item', 'gutenberg' ),
			url: '#placeholder',
```

I like these values better. 
- Url is a valid url now, meaning no validation problems. 
- Title matches the value of the default titleless custom menu item. 
- Title is transable, so if something goes wrong, the title is transable for the user.

If we need placeholder menu items, maybe we could add a field of post meta to the menu item, like is_placeholder and return this in the REST API, thoughts?

Reply:
Good call!